src/magic_spell_effect.cpp

```diff
@@ -116,9 +116,45 @@ static void build_line( spell_detail::line_iterable line, const tripoint &source
 }
 } // namespace spell_detail
 
-void spell_effect::teleport_random( const spell &sp, Creature &caster, const tripoint & )
+static bool in_spell_aoe( const tripoint &start, const tripoint &end, const int &radius,
+                          const bool ignore_walls )
+{
+    if( rl_dist( start, end ) > radius ) {
+        return false;
+    }
+    if( ignore_walls ) {
+        return true;
+    }
+    map &here = get_map();
+    const std::vector<tripoint> trajectory = line_to( start, end );
+    for( const tripoint &pt : trajectory ) {
+        if( here.impassable( pt ) ) {
+            return false;
+        }
+    }
+    return true;
+}
+
+void spell_effect::short_range_teleport( const spell &sp, Creature &caster, const tripoint &target )
 {
     bool safe = !sp.has_flag( spell_flag::UNSAFE_TELEPORT );
+    bool target_teleport = sp.has_flag( spell_flag::TARGET_TELEPORT );
+    if( target_teleport ) {
+        if( sp.aoe() == 0 ) {
+            teleport::teleport_to_point( caster, target, safe, false );
+            return;
+        }
```

Comment:
this isn't actually necessary. it might be more performant, but since this involves a character action i would prefer readability. so you can basically drop this, and if random_entry does what i think it does, if it only has one tripoint in the set it'll call target on the right point.